Have MediaDescription stop using AtomString

jyavenard opened the pull request:
#### 8904da90987b4e0a99d4ea716053ee65a613e15b
<pre>
Have MediaDescription stop using AtomString
<a href="https://bugs.webkit.org/show_bug.cgi?id=268728">https://bugs.webkit.org/show_bug.cgi?id=268728</a>
<a href="https://rdar.apple.com/122292724">rdar://122292724</a>

Reviewed by Youenn Fablet.

Sub-task of 265982.

To be used in WorkQueues, MediaDescription needs to stop using AtomString.
Instead we make this object use ThreadSafeRefCount, and return a StringView
which explicitly enforce that you can&apos;t extend the lifetime of the string
outside the lifetime of the MediaDescription container.

No change in observable behaviour.

* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment):
(WebCore::SourceBuffer::validateInitializationSegment):
* Source/WebCore/platform/MediaDescription.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.mm:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
* Source/WebCore/platform/graphics/gstreamer/mse/GStreamerMediaDescription.cpp:
(WebCore::GStreamerMediaDescription::codec const):
(WebCore::GStreamerMediaDescription::extractCodecName const):
(WebCore::GStreamerMediaDescription::extractCodecName): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/GStreamerMediaDescription.h:
(WebCore::GStreamerMediaDescription::GStreamerMediaDescription):
* Source/WebCore/platform/mock/mediasource/MockBox.h:
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.cpp:
* Source/WebKit/GPUProcess/media/MediaDescriptionInfo.h:
(WebKit::MediaDescriptionInfo::MediaDescriptionInfo):
* Source/WebKit/GPUProcess/media/MediaDescriptionInfo.serialization.in:
* Source/WebKit/GPUProcess/media/RemoteMediaDescription.h:
(WebKit::RemoteMediaDescription::RemoteMediaDescription):

Canonical link: <a href="https://commits.webkit.org/274178@main">https://commits.webkit.org/274178@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa57b851f2273d98ac01b385253a28adbdcbeb91

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38073 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16984 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40342 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40620 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33861 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40124 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19704 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14319 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32150 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38647 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14342 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33331 "Build was cancelled. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Skipping applying patch since patch_id isn't provided; Checked out pull request") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12506 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12437 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34062 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41896 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34561 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34512 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38322 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13044 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10708 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36514 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14593 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8558 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13457 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14043 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
```diff
@@ -489,7 +489,7 @@ class MediaDescriptionWebM final : public MediaDescription {
 
 private:
-    mutable std::optional<AtomString> m_codec;
+    mutable std::optional<String> m_codec;
```
Why an optional here.
Can't we just use isNull which represents a null RefPtr?

jyavenard replied:
values are lazily constructed because it's using AtomString. So internally, those objets either use a std::string or a NSString ; it is assume that the thread on which the object is read, is the only one that will ever be read.

But that assumption will not hold for long.

What I will do is always generate the string on the fly then, so you get a new string.

Using a non-thread safe refcount isn't a workable solution

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.mm:
```diff
@@ -169,16 +169,9 @@ - (void)streamDataParser:(AVStreamDataParser *)streamDataParser didProvideConten
     bool isAudio() const override { return m_isAudio; }
     bool isText() const override { return m_isText; }
-    AtomString codec() const override
+    String codec() const override
```
final maybe

youennf commented on Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.cpp:
```diff
@@ -108,7 +108,7 @@ class MockMediaDescription final : public MediaDescription {
     virtual ~MockMediaDescription() = default;
 
-    AtomString codec() const override { return m_box.codec(); }
+    String codec() const override { return m_box.codec(); }
```
final

youennf commented on Source/WebCore/platform/MediaDescription.h:
```diff
@@ -27,15 +27,15 @@
     virtual ~MediaDescription() = default;
 
-    virtual AtomString codec() const = 0;
+    virtual String codec() const = 0;
```
It seems fragile to have a String (a RefCounted), owned by a ThreadSafeRefCounted.
I would be tempted to keep a RefCounted here, and do explicit steps when hopping this data to another thread.

jyavenard replied:
I disagree. There's plenty of thread-safe refcounted object that contains a String. It's the container that is changed, not the the string itself.

it's just a plain container. I could change the method the be returning a `const String&` instead if that makes it better.

But I don't see how you can transfer this object across thread, if all of a sudden all we can do is move it (which is what your comment implies)

jyavenard replied:
Ok, so I made the codec method now return a StringView ; it makes it more explicit about the lifetime of the object and also not possible to take a reference to the inner's String.

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaDescription.h:
```diff
@@ -40,21 +40,21 @@ class RemoteMediaDescription : public WebCore::MediaDescription {
 private:
     RemoteMediaDescription(const MediaDescriptionInfo& descriptionInfo)
-        : m_codec(descriptionInfo.m_codec)
+        : m_codec(descriptionInfo.m_codec.isolatedCopy())
```
This seems a bit fishy. Can we instead keep the object RefCounted and do an isolatedCopy() when we send this object to another thread?
Is it only RemoteMediaDescription that needs to be hopped to another thread?

jyavenard replied:
The MediaDescription is already jumping across threads/process via a reference

but is read in the main thread.

Following https://github.com/WebKit/WebKit/pull/23850 it has to be copied as the SourceBufferPrivateRemote now does hop threads.

jyavenard replied:
BTW, isolatedCopy on the String that comes from IPC is a no-op, as they are already isolated

youennf replied:
WTFMove(descriptionInfo.m_codec).isolatedCopy() is optimised, descriptionInfo.m_codec.isolatedCopy() is doing a memory allocation.

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.mm:
```diff
@@ -194,11 +187,12 @@ AtomString codec() const override
 
     FourCC m_originalCodec;
-    mutable std::optional<AtomString> m_codec;
+    String m_codec;
```
Could be const with a small refactoring

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaDescription.h:
```diff
@@ -40,21 +40,21 @@ class RemoteMediaDescription : public WebCore::MediaDescription {
     }
 
-    AtomString m_codec;
+    String m_codec;
```
const

jyavenard replied:
ok, I've refactored to make m_codec a member of the base class.

youennf reviewed (approved):
I would be tempted to move the `const String m_codec` and StringView getter to MediaDescription (with an ASSERT(m_codec.isSafeToSendToAnotherThread() in MediaDescription constructor).
That way, the model is defined in a single place, instead of each implementation to think about it.

jyavenard committed:
Have MediaDescription stop using AtomString
https://bugs.webkit.org/show_bug.cgi?id=268728
rdar://122292724

Reviewed by Youenn Fablet.

Sub-task of 265982.

To be used in WorkQueues, MediaDescription needs to stop using AtomString.
Instead we make this object use ThreadSafeRefCount, and return a StringView
which explicitly enforce that you can't extend the lifetime of the string
outside the lifetime of the MediaDescription container.

No change in observable behaviour.

* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment):
(WebCore::SourceBuffer::validateInitializationSegment):
* Source/WebCore/platform/MediaDescription.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferParserAVFObjC.mm:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
* Source/WebCore/platform/graphics/gstreamer/mse/GStreamerMediaDescription.cpp:
(WebCore::GStreamerMediaDescription::codec const):
(WebCore::GStreamerMediaDescription::extractCodecName const):
(WebCore::GStreamerMediaDescription::extractCodecName): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/GStreamerMediaDescription.h:
(WebCore::GStreamerMediaDescription::GStreamerMediaDescription):
* Source/WebCore/platform/mock/mediasource/MockBox.h:
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.cpp:
* Source/WebKit/GPUProcess/media/MediaDescriptionInfo.h:
(WebKit::MediaDescriptionInfo::MediaDescriptionInfo):
* Source/WebKit/GPUProcess/media/MediaDescriptionInfo.serialization.in:
* Source/WebKit/GPUProcess/media/RemoteMediaDescription.h:
(WebKit::RemoteMediaDescription::RemoteMediaDescription):

Canonical link: https://commits.webkit.org/274178@main
